server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,101 @@
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link JsTypeRegistration} for Jetty.
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsTypeRegistration {
+
+    /**
+     * Creates a new js plugins instance with a temporary directory.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final JsPlugins jsPlugins = new JsPlugins(Files.createTempDirectory(JsPlugins.class.getName()));
+        jsPlugins.writeManifest(); // empty manifest
+        return jsPlugins;
+    }
+
+    private final Path path;
+    private final Map<String, JsType> plugins;
+
+    private JsPlugins(Path path) {
+        this.path = Objects.requireNonNull(path);
+        this.plugins = new LinkedHashMap<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", path.toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "true");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsType jsType) {
+        if (plugins.containsKey(jsType.name())) {
+            throw new IllegalArgumentException(
+                    String.format("js plugin with name '%s' already exists", jsType.name()));
+        }
+        try {
+            jsType.copyTo(path.resolve(jsType.name()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        plugins.put(jsType.name(), jsType);
+        try {
+            writeManifest();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeManifest() throws IOException {
+        // note: would this be better as a custom servlet instead? (this current way is certainly easy...)
+        try (final OutputStream out = new BufferedOutputStream(Files.newOutputStream(path.resolve("manifest.json")))) {
+            final List<Map<String, String>> pluginsList = new ArrayList<>(plugins.size());
+            for (JsType value : plugins.values()) {
+                final Map<String, String> p = new LinkedHashMap<>(3);
+                p.put("name", value.name());
+                p.put("version", value.version());
+                p.put("main", value.main());
```

Comment:
The JS side right now assumes the path is relative to `js-plugins/<name>`. This is the convention that @mofojed has established - but I'm fine to change it if we want.

Happy to add validation for name; are all NPM package names url safe? Note, the names _do_ contain `@` symbol and that works ok.

Reply:
slashes are legal too iirc, but i dont think we want to encourage subdirectories? Also, since we want to allow non-npm modules, we might need to handle collisions, perhaps by generating dir names.

Reply:
Well, the deephaven plugins _do_ include a slash for subdirectory implications: https://www.npmjs.com/search?q=keywords%3Adeephaven-js-plugin

* [@deephaven/js-plugin-dashboard-object-viewer](https://www.npmjs.com/package/@deephaven/js-plugin-dashboard-object-viewer)
* [@deephaven/js-plugin-table-example](https://www.npmjs.com/package/@deephaven/js-plugin-table-example)
* [@deephaven/js-plugin-matplotlib](https://www.npmjs.com/package/@deephaven/js-plugin-matplotlib)
* [@deephaven/js-plugin-plotly](https://www.npmjs.com/package/@deephaven/js-plugin-plotly)

Reply:
I'm not sure dealing with collisions is something I'm practically concerned with right now. We _could_ generate random prefixes (or full paths) to deal with that, but it seems like overkill ATM.

How does js plugin resolution work? For example, let's say we accidentally installed two separate versions of the js-plugin-plotly NPM package (and we "allowed" this by distinguishing them with random prefixes), how would the web UI handle this?

@mofojed

Reply:
I am not concerned about multiple versions of same package right now. If we need that, can do aliasing.

The layout is similar to how `node_modules` is laid out right now - the subdirectory is intentional, as then all `@deephaven` plugins are in the same dir ([scope](https://docs.npmjs.com/cli/v8/using-npm/scope))